### PR TITLE
fix(rome_js_parser): jsx vs type assertions

### PR DESCRIPTION
**Pull request by xunilrj**
This PR fixes a problem with jsx versus type assertions at primary expressions.

Currently, something like the code below was failing:

```
let a = <number>b;
```

We would treat ```b``` as ```JSXText``` and later complaint about a missing closing tag.
Now we do our best to parse JSX and type assertions separately and warn appropriately.

This changes an error on an HTML comment test, which does not matter because we still do not support these comments.
If anyone thinks that the current message is worst, I can try something to improve it.

### CODE REVIEW

**Comment by xunilrj** [comment]
About the regressions:

#### compiler/newExpressionWithCast.ts
We were incorrectly accepting this file. It does not have an associated ```.errors.txt``` to signal parsing errors, but it does contain parsing errors. So I think we are correct now. Specifically this part:

```ts
// parse error
var test2 = new <any>Test2();
```

#### conformance/parser/ecmascript5/Expressions/parserTypeAssertionInObjectCreationExpression1.ts

Exact same thing:

```
new <T>Foo()
```

**Comment by xunilrj** [comment]
!bench_parser

**Review by ematipico — COMMENTED** [review]
Looks good to me!

**Review by ematipico — APPROVED** [review]

File: crates/rome_js_parser/src/syntax/jsx/mod.rs [hunk 1]
```diff
@@ -103,8 +107,12 @@ fn parse_any_jsx_tag(p: &mut Parser, in_expression: bool) -> ParsedSyntax {
...
             parse_jsx_children(p);
 
-            expect_closing_element(p, in_expression, name, opening_range);
-            Present(element.complete(p, JSX_ELEMENT))
+            if expect_closing_element(p, in_expression, name, opening_range).is_none() {
+                element.abandon(p);
+                Absent
+            } else {
+                Present(element.complete(p, JSX_ELEMENT))
+            }
```
Comment (MichaReiser) [inline crates/rome_js_parser/src/syntax/jsx/mod.rs:115]: This is dangerous. It will progress the parser but return `Absent`. The processed tokens now get lost if the caller isn't using a checkpoint. Yes, you added a checkpoint inside of `parse_expression` but this function is also called when parsing the children of an element, in which case the parser silently skips over content.

File: crates/rome_js_parser/src/syntax/jsx/mod.rs [hunk 2]
```diff
@@ -213,10 +221,18 @@ fn expect_closing_element(
...
+
+    if !p.expect(T![<]) {
+        m.abandon(p);
+        return None;
+    }
+
+    if !p.expect(T![/]) {
+        m.abandon(p);
+        return None;
+    }
```
Comment (MichaReiser) [inline crates/rome_js_parser/src/syntax/jsx/mod.rs:235]: Same here. The parser has now eaten over at least one `<` and then returns `Absent`... which violates the constraint that parse methods should never progress the parser if they return `Absent`.

File: crates/rome_js_parser/test_data/inline/ok/type_assertion_primary_expression.ts [hunk 3]
```diff
@@ -0,0 +1 @@
+let a = <number>undefined;
```
Comment (MichaReiser) [inline crates/rome_js_parser/test_data/inline/ok/type_assertion_primary_expression.ts:1]: This case works for me even without this fix in place. I'm going to revert this change. The slightly less good error message is tolerable in my view. We can add explicit handling of `let a = new<number> Test()` if we think this is a common pattern but we need to change the error message to mention that type assertion are disallowed in case JSX is enabled.
